### PR TITLE
Implemented laravel spatie permission package and role crud

### PR DESCRIPTION
**Pull request by ankitInfyOm**

### CODE REVIEW

File: app/Http/Controllers/RoleController.php [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+     * @param Role $role
+     *
+     * @throws Exception
+     *
+     * @return JsonResponse|RedirectResponse|Redirector
+     */
+    public function destroy(Role $role)
+    {
+        if ($role->users()->count() > 0) {
+            throw new BadRequestHttpException(
```
Comment (vishalinfyom) [inline app/Http/Controllers/RoleController.php:152]: this exception should be `UnprocessableEntrity`

File: app/Repositories/UserRepository.php [hunk 2]
```diff
@@ -61,7 +61,9 @@ public function store($input)
                 $input['image_path'] = ImageTrait::makeImage($input['photo'], User::IMAGE_PATH);
             }
 
-            User::create($input);
+            $user =User::create($input);
```
Comment (vishalinfyom) [inline app/Repositories/UserRepository.php]: add space after `=`

File: database/seeds/PermissionTableSeeder.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+     *
+     * @return void
+     */
+    public function run()
+    {
+        $input =
+            [
+                'name'         => 'manage_users',
+                'display_name' => 'Manage Users',
+                'description'  => 'User tab visible',
```
Comment (vishalinfyom) [inline database/seeds/PermissionTableSeeder.php:19]: Can Manage Users

File: resources/views/roles/create.blade.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
+@extends('layouts.app')
+
```
Comment (vishalinfyom) [inline resources/views/roles/create.blade.php]: add title everywhere

File: resources/views/users/index.blade.php [hunk 5]
```diff
@@ -92,7 +92,7 @@
 
       // open delete confirmation model
       function deleteData(id) {
-        deleteItem('{{url('users')}}/' + id, '#users-table', 'User');
+        deleteItem('{{url('users')}}/' + id, '#users_table', 'User');
```
Comment (vishalinfyom) [inline resources/views/users/index.blade.php]: id should be `tblUsers`

File: resources/views/roles/edit_fields.blade.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        {!! Form::label('description', 'Description:') !!}
+        {!! Form::textarea('description', null, ['class' => 'form-control','rows'=>5]) !!}
+    </div>
+</div>
+<div class="col-6 pl-0">
+    <div class="form-group col-sm-12 col-lg-12 pl-0 pr-0">
+        {!! Form::label('permissions', 'Permissions:') !!}
+        <div class="row">
+            @foreach($permissions as $key=>$value)
+                @if(in_array($key,$roles->permissions->pluck('id')->toArray()))
```
Comment (vishalinfyom) [inline resources/views/roles/edit_fields.blade.php:20]: need to refactor this

**Review by vishalinfyom — CHANGES_REQUESTED** [review]

File: app/Http/Requests/UpdateRoleRequest.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        $rules['name'] = 'required|unique:roles,name,'.$this->route('role')->id;
```
Comment (ajayinfyom) [inline app/Http/Requests/UpdateRoleRequest.php:29]: here, need to override existing rules from the role model.
```
$rules = Role::$rules;
$rules['name'] = 'required|unique:roles,name,'.$this->route('role')->id; 
```

**Review by ajayinfyom — COMMENTED** [review]
- Remove `composer.lock` file.
- Need to merge the latest code.

File: app/Http/Controllers/RoleController.php [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+     * @param Role $role
+     *
+     * @return Response
+     */
+    public function edit(Role $role)
+    {
+        /** @var Permission $permissions */
+        $permissions = $this->permissionRepository->permissionList();
+
+        return view('roles.edit')->with(['roles' => $role, 'permissions' => $permissions]);
```
Comment (ajayinfyom) [inline app/Http/Controllers/RoleController.php:122]: its should be `role`.

File: app/Http/Controllers/UserController.php [hunk 9]
```diff
@@ -116,8 +123,10 @@ public function edit($id, Request $request)
 
             return redirect(route('users.index'));
         }
+        $roles = $this->roleRepo->getRolesList();
+        $selectedRoles = $user->roles()->pluck('role_id')->toArray();
```
Comment (ajayinfyom) [inline app/Http/Controllers/UserController.php:127]: change `$selectedRoles` to `$selectedRole`

File: resources/views/users/edit_fields.blade.php [hunk 10]
```diff
@@ -28,6 +28,10 @@
              src="{{asset($user->image_path)}}"/>
     </div>
 </div>
+<div class="form-group col-sm-12">
+    {!! Form::label('role_id', 'Role') !!} <span class="required">*</span>
+    {!! Form::select('role_ids[]',$roles, $selectedRoles, ['class' => 'form-control','id'=>'editRoleId','multiple'=>true,'required']) !!}
```
Comment (ajayinfyom) [inline resources/views/users/edit_fields.blade.php:33]: change  `$selectedRoles` to `$selectedRole`

File: app/Http/Controllers/RoleController.php [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+    /**
+     * RoleController constructor.
+     *
+     * @param RoleRepository       $rolesRepo
+     * @param PermissionRepository $permissionRepository
+     */
+    public function __construct(RoleRepository $rolesRepo, PermissionRepository $permissionRepository)
+    {
+        $this->rolesRepository = $rolesRepo;
```
Comment (ajayinfyom) [inline app/Http/Controllers/RoleController.php:40]: change to `$rolesRepo` to `$roleRepository`

**Review by ajayinfyom — CHANGES_REQUESTED** [review]
* Also need to be fixed against to Laravel 7.0

**Comment by ShaileshInfyom** [comment]
no longer valid

**Pull request closed without merging**